### PR TITLE
{devel}[fosscuda/2020b] PyTorch v1.7.1 w/ Python 3.8.6

### PR DESCRIPTION
**Pull request by branfosj**
(created using `eb --new-pr`)

Notes: 
* I've used `cuDNN-8.0.4.30-CUDA-11.1.1.eb` as there does not seem to be a `cuDNN` / `8.0.5.39` / `CUDA` `11.1` for ppc64le
* I've dropped the `PyTorch-1.7.1_el8_ppc64le.patch`, compared to the `fosscuda/2019b` install, as this should only be needed on RHEL 8 with CUDA <= 10

Patches added through #12147 
* `PyTorch-1.7.1_validate-num-gpus-in-distributed-test.patch` fixes the NCCL error seen during tests
* `PyTorch-1.7.1_complex32.patch` provides complex32 dtype
* `PyTorch-1.7.1_bypass-nan-compare.patch` disables two tests that are known to fail. PyTorch are working on a fix for these failures.

### CODE REVIEW

File: easybuild/easyconfigs/p/PyTorch/PyTorch-1.7.1-fosscuda-2020b.eb [hunk 1]
```diff
@@ -0,0 +1,314 @@
...
+    ('numactl', '2.0.13'),
+    ('FFmpeg', '4.3.1'),
+    ('Pillow', '8.0.1'),
+    ('cuDNN', '8.0.4.30', '-CUDA-11.1.1', True),
+    ('magma', '2.5.4'),
+    ('NCCL', '2.8.3', '-CUDA-11.1.1', True),
+]
+
+# default CUDA compute capabilities to use (override via --cuda-compute-capabilities)
+cuda_compute_capabilities = ['3.5', '3.7', '5.2', '6.0', '6.1', '7.0', '7.2', '7.5']
```
Comment (Micket) [inline easybuild/easyconfigs/p/PyTorch/PyTorch-1.7.1-fosscuda-2020b.eb:37]:
```suggestion
cuda_compute_capabilities = ['3.5', '3.7', '5.2', '6.0', '6.1', '7.0', '7.2', '7.5', '8.0']
```

**Comment by branfosj** [comment]
Test report by @branfosj
**FAILED**
Build succeeded for 0 out of 1 (2 easyconfigs in total)
bear-pg0212u15b.bear.cluster - Linux centos linux 8.2.2004, x86_64, Intel(R) Xeon(R) CPU E5-2640 v4 @ 2.40GHz (broadwell), Python 3.6.8
See https://gist.github.com/0ed5717dff765a4e7c1b622fdf09ceb5 for a full test report.

**Comment by branfosj** [comment]
The test log shows:
```
======================================================================
FAIL: test_reduce_sum_cuda (__main__.TestDistBackendWithFork)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/scratch/branfosj-admin/eb-11ozam_e/tmp8lgl4_dn/lib/python3.8/site-packages/torch/testing/_internal/common_distributed.py", line 267, in wrapper
    self._join_processes(fn)
  File "/scratch/branfosj-admin/eb-11ozam_e/tmp8lgl4_dn/lib/python3.8/site-packages/torch/testing/_internal/common_distributed.py", line 384, in _join_processes
    self._check_return_codes(elapsed_time)
  File "/scratch/branfosj-admin/eb-11ozam_e/tmp8lgl4_dn/lib/python3.8/site-packages/torch/testing/_internal/common_distributed.py", line 436, in _check_return_codes
    self.assertEqual(
  File "/scratch/branfosj-admin/eb-11ozam_e/tmp8lgl4_dn/lib/python3.8/site-packages/torch/testing/_internal/common_utils.py", line 1173, in assertEqual
    self.assertTrue(result, msg=msg)
AssertionError: False is not true : Expected zero exit code but got 1
```

But before that we see
```
test_reduce_sum_cuda (__main__.TestDistBackendWithFork) ... Process process 1:
Process process 0:
Process process 2:
Traceback (most recent call last):
  File "/rds/bear-apps/devel/eb-sjb-up/EL8/EL8-has/software/Python/3.8.6-GCCcore-10.2.0/lib/python3.8/multiprocessing/process.py", line 315, in _bootstrap
    self.run()
  File "/rds/bear-apps/devel/eb-sjb-up/EL8/EL8-has/software/Python/3.8.6-GCCcore-10.2.0/lib/python3.8/multiprocessing/process.py", line 108, in run
    self._target(*self._args, **self._kwargs)
  File "/scratch/branfosj-admin/eb-11ozam_e/tmp8lgl4_dn/lib/python3.8/site-packages/torch/testing/_internal/distributed/distributed_test.py", line 294, in _run
    dist.init_process_group(
  File "/scratch/branfosj-admin/eb-11ozam_e/tmp8lgl4_dn/lib/python3.8/site-packages/torch/distributed/distributed_c10d.py", line 455, in init_process_group
    barrier()
  File "/scratch/branfosj-admin/eb-11ozam_e/tmp8lgl4_dn/lib/python3.8/site-packages/torch/distributed/distributed_c10d.py", line 1960, in barrier
    work = _default_pg.barrier()
RuntimeError: NCCL error in: ../torch/lib/c10d/ProcessGroupNCCL.cpp:784, invalid usage, NCCL version 2.8.3
Traceback (most recent call last):
  File "/rds/bear-apps/devel/eb-sjb-up/EL8/EL8-has/software/Python/3.8.6-GCCcore-10.2.0/lib/python3.8/multiprocessing/process.py", line 315, in _bootstrap
    self.run()
  File "/rds/bear-apps/devel/eb-sjb-up/EL8/EL8-has/software/Python/3.8.6-GCCcore-10.2.0/lib/python3.8/multiprocessing/process.py", line 108, in run
    self._target(*self._args, **self._kwargs)
  File "/scratch/branfosj-admin/eb-11ozam_e/tmp8lgl4_dn/lib/python3.8/site-packages/torch/testing/_internal/distributed/distributed_test.py", line 294, in _run
    dist.init_process_group(
  File "/scratch/branfosj-admin/eb-11ozam_e/tmp8lgl4_dn/lib/python3.8/site-packages/torch/distributed/distributed_c10d.py", line 455, in init_process_group
    barrier()
  File "/scratch/branfosj-admin/eb-11ozam_e/tmp8lgl4_dn/lib/python3.8/site-packages/torch/distributed/distributed_c10d.py", line 1960, in barrier
    work = _default_pg.barrier()
RuntimeError: NCCL error in: ../torch/lib/c10d/ProcessGroupNCCL.cpp:784, invalid usage, NCCL version 2.8.3
Traceback (most recent call last):
  File "/rds/bear-apps/devel/eb-sjb-up/EL8/EL8-has/software/Python/3.8.6-GCCcore-10.2.0/lib/python3.8/multiprocessing/process.py", line 315, in _bootstrap
    self.run()
  File "/rds/bear-apps/devel/eb-sjb-up/EL8/EL8-has/software/Python/3.8.6-GCCcore-10.2.0/lib/python3.8/multiprocessing/process.py", line 108, in run
    self._target(*self._args, **self._kwargs)
  File "/scratch/branfosj-admin/eb-11ozam_e/tmp8lgl4_dn/lib/python3.8/site-packages/torch/testing/_internal/distributed/distributed_test.py", line 294, in _run
    dist.init_process_group(
  File "/scratch/branfosj-admin/eb-11ozam_e/tmp8lgl4_dn/lib/python3.8/site-packages/torch/distributed/distributed_c10d.py", line 455, in init_process_group
  File "/scratch/branfosj-admin/eb-11ozam_e/tmp8lgl4_dn/lib/python3.8/site-packages/torch/distributed/distributed_c10d.py", line 455, in init_process_group
    barrier()
  File "/scratch/branfosj-admin/eb-11ozam_e/tmp8lgl4_dn/lib/python3.8/site-packages/torch/distributed/distributed_c10d.py", line 1960, in barrier
    work = _default_pg.barrier()
RuntimeError: NCCL error in: ../torch/lib/c10d/ProcessGroupNCCL.cpp:784, invalid usage, NCCL version 2.8.3
FAIL
```

**Comment by branfosj** [comment]
As expected, from looking at bug reports, this error persists with NCCL 2.7.8:
```
RuntimeError: NCCL error in: ../torch/lib/c10d/ProcessGroupNCCL.cpp:784, invalid usage, NCCL version 2.7.8
```

**Comment by branfosj** [comment]
Test report by @branfosj
**FAILED**
Build succeeded for 1 out of 2 (2 easyconfigs in total)
bear-pg0212u15b.bear.cluster - Linux centos linux 8.2.2004, x86_64, Intel(R) Xeon(R) CPU E5-2640 v4 @ 2.40GHz (broadwell), Python 3.6.8
See https://gist.github.com/5a3089581e3d0cc2b1ce0da0075360de for a full test report.

**Comment by branfosj** [comment]
Test report by @branfosj
**FAILED**
Build succeeded for 1 out of 2 (2 easyconfigs in total)
bear-pg0212u15b.bear.cluster - Linux centos linux 8.2.2004, x86_64, Intel(R) Xeon(R) CPU E5-2640 v4 @ 2.40GHz (broadwell), Python 3.6.8
See https://gist.github.com/acef361bb0af5f4ece2bf61036b4894e for a full test report.

**Comment by branfosj** [comment]
Test report by @branfosj
**FAILED**
Build succeeded for 1 out of 2 (2 easyconfigs in total)
bear-pg0212u15b.bear.cluster - Linux centos linux 8.2.2004, x86_64, Intel(R) Xeon(R) CPU E5-2640 v4 @ 2.40GHz (broadwell), Python 3.6.8
See https://gist.github.com/83ffe845bf9a662dfbbcb6857547d1ed for a full test report.

**Comment by branfosj** [comment]
Test report by @branfosj
**SUCCESS**
Build succeeded for 2 out of 2 (2 easyconfigs in total)
bear-pg0212u17a.bear.cluster - Linux centos linux 8.2.2004, x86_64, Intel(R) Xeon(R) CPU E5-2640 v4 @ 2.40GHz (broadwell), Python 3.6.8
See https://gist.github.com/e81c3cca332c834f97a9ce702fed2399 for a full test report.

**Comment by Flamefire** [comment]
Test report by @Flamefire
**FAILED**
Build succeeded for 1 out of 2 (2 easyconfigs in total)
taurusi8032 - Linux centos linux 7.9.2009, x86_64, AMD EPYC 7352 24-Core Processor, Python 2.7.5
See https://gist.github.com/8158a48ad183c69131185f2b16bfedc7 for a full test report.

**Comment by Flamefire** [comment]
```
[W CudaIPCTypes.cpp:22] Producer process has been terminated before all shared CUDA tensors released. See Note [Sharing CUDA tensors]
Traceback (most recent call last):
  File "run_test.py", line 745, in <module>
    main()
  File "run_test.py", line 728, in main
    raise RuntimeError(err_message)
RuntimeError: distributed/test_c10d_spawn failed! Received signal: SIGSEGV
```

To me this looks like static deinit fiasco. Tests seem to be fine though

File: easybuild/easyconfigs/t/typing-extensions/typing-extensions-3.7.4.3-fosscuda-2020b.eb [hunk 2]
```diff
@@ -0,0 +1,21 @@
+easyblock = 'PythonPackage'
```
Comment (Flamefire) [inline easybuild/easyconfigs/t/typing-extensions/typing-extensions-3.7.4.3-fosscuda-2020b.eb:1]: There already is a typing-extensions-3.7.4.3-GCCcore-10.2.0.eb in the EC repo, so we don't need this one

Comment (branfosj) [reply easybuild/easyconfigs/t/typing-extensions/typing-extensions-3.7.4.3-fosscuda-2020b.eb]: Thanks. I've deleted it.

**Comment by Flamefire** [comment]
@branfosj I backported your patches to 2019b and included mine to fix the aformentioned bug and another one: https://github.com/easybuilders/easybuild-easyconfigs/pull/12147

I'm currently running your EC with those patches on our A100 cluster and see if that succeeds. Feel free to add those patches here.

**Comment by boegel** [comment]
@branfosj #12147 is merged, so this PR should be synced accordingly by pulling in the missing patches?

**Comment by branfosj** [comment]
I've synced with `develop` and matched the patches from `PyTorch-1.7.1-fosscuda-2019b-Python-3.7.4.eb` - with the exception of `PyTorch-1.7.1_el8_ppc64le.patch` which is removed as it is not needed with CUDA 11.

**Comment by boegel** [comment]
@boegelbot please test @ generoso

**Comment by branfosj** [comment]
Test report by @branfosj
**SUCCESS**
Build succeeded for 1 out of 1 (1 easyconfigs in total)
bear-pg0212u17a.bear.cluster - Linux centos linux 8.2.2004, x86_64, Intel(R) Xeon(R) CPU E5-2640 v4 @ 2.40GHz (broadwell), Python 3.6.8
See https://gist.github.com/11639ce44e435e6f21ff00766f1e54d7 for a full test report.

**Review by boegel — APPROVED** [review]
lgtm

**Comment by boegel** [comment]
Test report by @boegel
**SUCCESS**
Build succeeded for 1 out of 1 (1 easyconfigs in total)
node3300.joltik.os - Linux centos linux 7.9.2009, x86_64, Intel(R) Xeon(R) Gold 6242 CPU @ 2.80GHz (cascadelake), Python 3.6.8
See https://gist.github.com/5fd4948c6a69b9be7560879e89e5beeb for a full test report.

**Comment by Flamefire** [comment]
Tested this on our A100 cluster and seeing multiple minor failures:
> AssertionError: False is not true : Tensors failed to compare as equal! With rtol=1.3e-06 and atol=1e-05, found 1 element(s) (out of 200) whose difference(s) exceeded the margin of error (including 0 nan comp
arisons). The greatest difference was 1.1809170246124268e-05 (-0.11461969465017319 vs. -0.11460788547992706), which occurred at index (3, 47).
exiting process with exit code: 10
E

I assume this is because we have 8 GPUs/node and PyTorch is not ready to handle that. Will open a bug report there and can add a patch here

**Comment by boegel** [comment]
Test report by @boegel
**SUCCESS**
Build succeeded for 1 out of 1 (1 easyconfigs in total)
node3501.doduo.os - Linux RHEL 8.2, x86_64, AMD EPYC 7552 48-Core Processor (zen2), Python 3.6.8
See https://gist.github.com/32f2066a992a94424838bbe4464667c0 for a full test report.

**Comment by boegel** [comment]
@Flamefire Should we let that block this PR, or do we handle that in a follow-up PR?

**Comment by Flamefire** [comment]
> @Flamefire Should we let that block this PR, or do we handle that in a follow-up PR?

@boegel Up to you. I'd fix this in here but as I'm likely the only one seeing the problem I could open the follow-up PR to avoid having to send patches via mail or so ;)
I have a patch ready and that solves that problem. I'll run a test build overnight to see if there is anything else

**Comment by Flamefire** [comment]
The patch made the tests pass on a few tests but I still have precision related failures. As PyTorch seems to work on other systems I'd assume this is specific to A100s/CC8.0 and not worth holding back this PR.
If I find a solution I'll open a follow-up PR

**Comment by boegel** [comment]
Going in, thanks @branfosj!

**Pull request merged**